### ISSUE TITLE
ENABLE_AAC_LATM is needed for enabling audio output for HE-AAC 

### ISSUE THREAD

**Issue by dr-m** [reufer/rpihddevice#20]
Recently, a new free-to-air DVB-T2 bouquet [was introduced in my region](https://www.digita.fi/en/etusivu/hd-transition/channel-number-change/). The video is H.264 and will play fine, but the PCM stereo output through HDMI would be silent. Unlike the previous DVB-T2 bouquet, the audio format is not AC3 48kHz stereo, but aac_latm (HE-AAC) 48 kHz stereo. It turns out that this audio format is supported, but not by default:
```diff
diff --git a/Makefile b/Makefile
index 8b43980..4c9659d 100644
--- a/Makefile
+++ b/Makefile
@@ -85,7 +85,7 @@ ifeq ($(DEBUG_OVGSTAT), 1)
     DEFINES += -DDEBUG_OVGSTAT
 endif
 
-ENABLE_AAC_LATM ?= 0
+ENABLE_AAC_LATM ?= 1
 ifeq ($(ENABLE_AAC_LATM), 1)
     DEFINES += -DENABLE_AAC_LATM
 endif
```
If I compile `rpihddevice` with the above patch, there will be a preprocessor warning, but everything seems to work, using `libavcodec` 59:
```
audio.c:32:2: warning: #warning "experimental AAC-LATM frame parser enabled, only 2ch/48kHz supported!" [-Wcpp]
```
Based on a very quick test (a few seconds), this seems to work for me. I thought that I would mention this in case someone else runs into this.

**Comment by reufer** [comment]
I've never tested AAC-LATM, but if it's working for you, I'm happy to merge a PR enabling it!

**Comment by dr-m** [comment]
Yes, it has been working for me in the past couple of weeks. My use is a bit sporadic, mostly focusing on recordings. Last weekend, I watched one recorded movie (1 to 2 hours) from this bouquet with lots of transmission errors, and there was no trouble with the audio.

**Comment by dr-m** [comment]
I see that `cRpiAudioDecoder::LatmCheck()` is hard-coding the number of channels and the sample rate. I think that it needs to be improved before the warning about the experimental status can be removed. In the FFmpeg source code, I found out that the 0x56e… corresponds to `LATM_HEADER` and `LATM_MASK`. I will have to study that code deeper to find out how the parameters are encoded.

**Comment by dr-m** [comment]
Based on some references that I was able to find so far, it does not look like the sampling rate or the number of audio channels would be present in any LATM or LOAS header.

Finally, I think that I found a useful reference https://nordig.org/wp-content/uploads/2016/03/NorDig-Unified-Requirements-ver.-3.2.1.pdf that is applicable to my region.
Page 80:
>NorDig IRDs supporting HE-AAC (and thereby also AAC-LC) shall be capable of:
>* decoding HE-AAC Version 1 at Level 2 at sampling rates of 48 kHz according to ETSI TS 101 154 [26].
>* decoding, including downmixing HE-AAC Version 1 at Level 4 (multichannel, up to 5.1) at sampling rates of 48 kHz according to ETSI TS 101 154 [26], Annex C (downmix).
>* Re-encoding HE-AAC Version 1 at Level 4 (multichannel, up to 5.1) at sampling rates of 48 kHz according to TS 101 154 [26], Annex H to AC-3 or DTS.

This seems to suggest that hard-coding the sampling rate at 48 kHz could be acceptable. But, hard-coding to 2 channels is not.

Section 12.6.7 AAC descriptor (pages 163 and 164) includes two interesting tables:
## Table 12.28 AAC descriptor’s Profile and level values for NorDig IRDs.
| profile and level value | Description | NorDig comment |
| ---- | ---- | ---- |
| 0x51 | LC-AAC Level 2 | no SBR used, up to stereo (mono or stereo)
| 0x52 | LC-AAC Level 4 | no SBR used, up to 5.1 (e.g. mono, stereo, 5.1)
| 0x58 | HE-AAC Level 2 | SBR may be used, up to stereo
| 0x5A | HE-AAC Level 4 | SBR may be used, up to 5.1(e.g. mono, stereo, 5.1)
## Table 12.29 AAC descriptor’s AAC type values for NorDig IRDs.
| AAC type value | Description | NorDig comment |
| ---- | ---- | ---- |
| 0x01 | HE-AAC audio, single mono channel | trigger for normal (mono) audio |
| 0x03 | HE-AAC audio, stereo | trigger for normal (stereo) audio |
| 0x05 | HE-AAC audio, surround sound | trigger for normal (multichannel) audio |
| 0x42 | HE-AAC Receiver mixed Supplementary Audio as per annex E of TS 101 154 | If no supplementary audio descriptor is included, then this shall be trigger for receiver mixed Supplementary Audio and this audio stream may include any AD_descriptor in PES_private_data for pan and fade control. |
| 0x47 | HE-AAC receiver mix Audio Description for the visually impaired | If no supplementary audio descriptor is included, then this shall be trigger for receiver mixed Supplementary Audio and this audio stream may will not include any AD_descriptor in PES_private_data. |
| 0x48 | HE-AAC broadcaster mix | Audio Description for the visually impaired | If no supplementary audio descriptor is included, then this shall be trigger for broadcast pre-mixed Supplementary Audio |

Right after the table there is the following note:
>As stated in section 6.5.4.3, for AAC audio streams without AAC descriptors shall be assumed to be of AAC type 0x03 (level 2, Normal stereo).

This suggests that the hard-coded default (48 kHz, 2 channels) is reasonable, and might in fact work all the time when using rpihddevice with recordings of or live transmission from free-to-air channels. I could imagine multichannel audio to be available in some pay-TV movies.

Do the tables ring any bell to you? With some `git grep` I found the following in `cEvent::FixEpgBugs()` in the VDR source code:
```c++
                if (!p->description) {
                   switch (p->type) {
                     case 0x05: p->description = strdup("Dolby Digital"); break;
                     default: ; // all others will just display the language
                     }
                   EpgBugFixStat(11, ChannelID());
                   }
```
Is that just a coincidence, or could that 0x05 be the same as in Table 12.29?

**Comment by reufer** [comment]
Thanks a lot for your investigations! I did not invested a lot of time in HE-ACC since there was no recording nor broadcast to test, so no bell ringing...
But looking at your results, I'd be fine to enable HE-ACC with current hard coded parameters - the preprocessor warning may still indicate the fixed settings.

**Comment by dr-m** [comment]
[I investigated this some more](https://www.linuxtv.org/mailman3/hyperkitty/list/vdr@linuxtv.org/thread/FXLY5B4ZJTDMT56PHQXSWC6MKXS3MM2A/), including checking with https://rr-project.org where exactly `ffprobe` would detect the format of [a 5.1 audio sample](https://www.linuxtv.org/mailman3/hyperkitty/list/vdr@linuxtv.org/message/WEFUKGEJWDYVVYQT2IQ54J3HZNEGTSTF/attachment/3/AAC51channel-sample.mp4). I think that the next step would be to identify the exact bytes in the file that need to be examined, and then figure out where VDR and rpihddevice would be accessing them when playing back this sample as a recording (`00001.ts`). Unfortunately, I don’t think I can do that any time soon.
